Updates to the kernel messaging spec

jasongrout opened the pull request:
These came out of conversations with @minrk.

CC also @SylvainCorlay, who I think has been advocating for the deprecation of sending a shutdown message on the shell channel.

Marking as work-in-progress since I plan to work on this more in the near future. But I also think these changes are pretty self-contained, so I wouldn't regret us merging these now and submitting another PR later either.

SylvainCorlay commented:
cc @JohanMabille

rgbkrk commented on docs/messaging.rst:
```diff
@@ -122,6 +122,14 @@ A message is defined by the following four-dictionary structure::
+all clients connected to a kernel and should be constant over the lifetime of
+the client. A kernel session value, in message headers from a kernel, should be
+generated on kernel startup or restart and should be constant for the lifetime
+of the kernel.
```
Thanks for clarifying this one in the docs!

rgbkrk commented on docs/messaging.rst:
```diff
@@ -959,6 +967,12 @@ Message type: ``shutdown_reply``::
 
+.. versionchanged:: 5.4
+
+    Sending a ``shutdown_request`` message on the ``shell`` channel is deprecated.
```
What's the strategy for deprecating the `shutdown_request`?

We also deprecated payloads, yet we still use them and have no current replacement.

jasongrout replied:
Are you basically asking what deprecation means in this context? Good question. @minrk, thoughts?  Does this mean that if a kernel implements protocol 5.4, it doesn't need to listen to shutdown_request messages on the shell channel?

jasongrout replied:
Actually, that doesn't make much sense to me. How about this: we're just warning people now, and a kernel can officially ignore shutdown requests on the shell channel in spec 6.0 or later.

rgbkrk replied:
Works for me!

Is most of the intent behind this that the shutdown API from the server is used? Was shutdown request put in place to help ensure that background resources got cleaned up appropriately (like with a cluster, deleting all the executors / workers?)

minrk replied:
The notebook server uses shutdown_request on the control channel as a first, polite "please shutdown" before sending signals. This is deprecating the ~unused functionality of sending these messages on the shell channel. Does nteract send shutdown_request on either of these channels?

@jasongrout I notices that KernelClient.shutdown is sending it on the shell channel. This is a rarely, if ever used method, that I think would only be used by frontends that don't own the kernel (e.g. `qtconsole --existing`), so that should be updated.

SylvainCorlay replied:
I am +1 on disallowing the shutdown and interupt on shell.

I don't know of any kernel that do it anyways since our clients use the control channel for that.

Making this explicit is really important IMO.

jasongrout commented:
The part I deleted about the notebook kernel status messages I filled out into more detailed docs for interacting with a kernel through the notebook server: https://github.com/jupyter/notebook/pull/3765

jasongrout commented:
@SylvainCorlay and @JohanMabille - can you comment on why sending the shutdown message on the shell channel is a problem for you?

JohanMabille commented:
@jasongrout the client may want to restart the kernel while a request is running (and is possibly stalled). In that case, we want to be able to interrupt the execution of that request.

Having the guarantee that the shutdown message is sent on the control channel gives more flexibility when implementing the protocol.

jasongrout commented:
@JohanMabille - so we certainly agree that it usually makes more sense to send the shutdown request on the control channel, exactly for the message priority issues you mentioned.

> Having the guarantee that the shutdown message is sent on the control channel gives more flexibility when implementing the protocol

This is what we were looking for - How is it more flexible when there is the guarantee that the shutdown only comes on the control channel?

JohanMabille commented:
From my point of view the guarantee is not that the shutdown *only* comes on the control channel, it's just that it comes at least on the control channel, whether it is also sent on the shell channel or not does not really matter.

The flexibility (maybe simplicity is more appropriate) comes from the possibility to have one thread per channel, thus you can shutdown a kernel even if the shell thread is processing a request.

 With the current specification, you have to synchronize a thread that processes the requests with a thread that polls both control and shell socket.

jasongrout commented:
@minrk - please review the changes I just pushed. Opening a control channel for the shutdown message, and percolating it through the code, is actually much more invasive than I thought originally.

jasongrout commented on docs/messaging.rst:
```diff
@@ -1213,6 +1239,10 @@ it shall publish a status message with ``execution_state: 'idle'``.
+A kernel may send optional status messages with execution states other than
+`busy` or `idle`. For example, a kernel may send a status message with a
+`starting` execution state exactly once at process startup.
+
```
@minrk, do we know of other kernels publishing other status states? We made the 'starting' status optional because there *might* be a race condition preventing the client from ever seeing a starting message. However, it is nice if we have a universe of only three status states, rather than opening it up to any status the kernel might want to send.

As for the kernel status states that the notebook spoofs, we already agreed that those should come on a different channel, a notebook channel, so I think those status states shouldn't count here as optional states sent from a kernel.

jasongrout replied:
To be clear, I think I'm +1 on reverting this "optional status states" change, and just going back to mandating a 'starting' status, and having a defined universe of 'starting', 'busy', and 'idle' states sent from the kernel.

SylvainCorlay replied:
I don't have an opinion on this.

jasongrout commented:
> much more invasive than I thought originally.

In particular, the signature of some functions changed, so this might now be considered a backwards-incompatible change, depending on what we view as the public-facing API of the package.

SylvainCorlay commented on docs/messaging.rst:
```diff
@@ -122,9 +122,28 @@ A message is defined by the following four-dictionary structure::
+    The session id in a message header can be used to identify the sending entity.
+    For example, if a client disconnects and reconnects to a kernel, and messages
+    from the kernel have a different kernel session id than prior to the disconnect,
+    the client should assume that the kernel was restarted.
```
:+1:

SylvainCorlay commented on docs/messaging.rst:
```diff
@@ -934,8 +954,7 @@ multiple cases:
 reply message (which is otherwise empty), it can assume that the kernel has
-completed shutdown safely.  The request can be sent on either the `control` or
-`shell` channels.
+completed shutdown safely.  The request is sent on the `control` channel.
```
:+1: this is really important this is specified here. +1000

SylvainCorlay commented:
@jasongrout is this still a WIP PR?

jasongrout commented:
> @jasongrout is this still a WIP PR?

The WIP part is that we wanted to revert the part about optional status messages.

jasongrout commented:
I just pushed a commit reverting the status message changes. I think this is now ready to go and has addressed the concerns raised in the comment process over the last year.

SylvainCorlay commented:
Awesome. I am :+1: for mergin now.

Leaving this open for a bit more to give others the occasion to comment!

cc @minrk

SylvainCorlay commented:
Cc @johanmabille. This is in!
